extensions/ql-vscode/src/test-manager.ts

```diff
@@ -0,0 +1,371 @@
+import { readFile } from "fs-extra";
+import {
+  CancellationToken,
+  Location,
+  Range,
+  TestController,
+  TestItem,
+  TestMessage,
+  TestRun,
+  TestRunProfileKind,
+  TestRunRequest,
+  Uri,
+  WorkspaceFolder,
+  WorkspaceFoldersChangeEvent,
+  tests,
+  workspace,
+} from "vscode";
+import { DisposableObject } from "./pure/disposable-object";
+import {
+  QLTestDirectory,
+  QLTestDiscovery,
+  QLTestFile,
+  QLTestNode,
+} from "./qltest-discovery";
+import { CodeQLCliServer } from "./cli";
+import { getErrorMessage } from "./pure/helpers-pure";
+import { BaseLogger, LogOptions } from "./common";
+import { TestRunner } from "./test-runner";
+import { TestManagerBase } from "./test-manager-base";
+import { App } from "./common/app";
+
+/**
+ * Returns the complete text content of the specified file. If there is an error reading the file,
+ * an error message is added to `testMessages` and this function returns undefined.
+ */
+async function tryReadFileContents(
+  path: string,
+  testMessages: TestMessage[],
```

Comment:
I'd slightly prefer changing this function so that it actually throws the error message and the call site is wrapped in a try-catch that would add to the testMessages array. This is more stylistic, so what you have now is also fine.

Reply:
I only did it this way to avoid duplicating the code that formats and adds the test message. For a non-exported function, I think it's OK as-is.